### PR TITLE
icarous

### PR DESCRIPTION
**Pull request by auturgy**
This change re-structures the icarous mavlink xml to expose the messages to users of ardupilotmega.xml, rather than requiring GCS's use icarous.xml directly.  Intent is to aid use/uptake of the icarous tools https://github.com/nasa/icarous
ICAROUS (Integrated Configurable Algorithms for Reliable Operations of Unmanned Systems) is a software architecture that enables the robust integration of mission specific software modules and highly assured core software modules for building safety-centric autonomous unmanned aircraft applications. The set of core software modules includes formally verified algorithms that detect, monitor, and control conformance to safety criteria; avoid stationary obstacles and maintain a safe distance from other users of the airspace; and compute resolution and recovery maneuvers, autonomously executed by the autopilot, when safety criteria are violated or about to be violated.

The ICAROUS package is currently available in C++ and Java. Both versions of ICAROUS have the same functionalities. The default version of ICAROUS supports the MAVLink protocol. You can define your own interfaces for other protocols (See C++/src/INTERFACE for examples)

### CODE REVIEW

**Comment by tridge** [comment]
If possible it would be good to use the mavlink2 ID range (more than 255), perhaps reserving 100 or so IDs for the project to prevent collisions

**Comment by auturgy** [comment]
No worries.  Is there a specific range that would be best?  Response from NASA was positive, so I'd like to get this in.

**Comment by peterbarker** [comment]
@auturgy check the ranges currently used - in ArduPilot, PX4Firmware and Papparazi.

Just take a range is 500 or so above one already taken.  Maybe 42000?

File: message_definitions/v1.0/icarous.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	  </message>
+
+	  <message id="224" name="SAFEGUARD">
+	    <description>Safeguard input</description>
+	    <field type="int8_t" name="value">1-success,0-failure</field>
+	  </message>
+	  
+	  <message id="229" name="HEARTBEAT_ICAROUS">
+	    <description>ICAROUS heartbeat</description>
+	    <field type="int8_t" name="status">FMS state (0-idle, 1-Takeoff, 2-Climb, 3-Cruise, 4-Land)</field>
```
Comment (peterbarker) [inline message_definitions/v1.0/icarous.xml]: This should be an enumeration.

File: message_definitions/v1.0/icarous.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0"?>
+<mavlink>
+	<enums>
+	</enums>
+	<messages>
+	  <message id="220" name="KINEMATIC_BANDS">
+	    <description>Kinematic multi bands (track) output from Daidalus</description>
+	    <field type="int8_t" name="numBands">Number of track bands</field>
+	    <field type="int8_t" name="type1">type of band (0 - None, 1-NEAR, ...)</field>
```
Comment (peterbarker) [inline message_definitions/v1.0/icarous.xml]: This might want to be an enumeration.

**Comment by magicrub** [comment]
Note, this is diverging further from mavlink/mavlink. Does NASA know and/or care that in order to use this xml, they MUST use Ardupilot's xml?

**Comment by auturgy** [comment]
Yes. The NASA implementation uses icarous.xml as the top level dialect, with ardupilotmega.xml as an include.
The implementation in this PR retains the NASA messageID’s and reliance on ardupilotmega.xml, but demotes icarous.xml into ardupilotmega.xml so as to not break pymavlink/mavproxy/ardupilotmega consumers.
We need to do our best to not diverge from upstream common.xml but this PR doesn’t impact that.

Regards,

James

**Comment by magicrub** [comment]
I don't mind merging this, but they should *SERIOUSLY* consider moving to MAVLink 2 with ID >255 so we can assign them a whole range that they can do whatever that to with (like uAvionix)

**Comment by peterbarker** [comment]
On Fri, 8 Dec 2017, Tom Pittenger wrote:

> I don't mind merging this, but they should SERIOUSLY consider moving to
> MAVLink 2 with ID >255 so we can assign them a whole range that they can do
> whatever that to with (like uAvionix)

They've already said that's OK, I believe.  I've suggested a range.

**Comment by auturgy** [comment]
I’m in Iceland: if one of you two can roll the changes to Peter’s suggested range, awesome. Else it’s going to be two weeks.

Regards,

James

**Comment by auturgy** [comment]
I've pulled in Swee's new icarous.xml, but it's messy doing this through a browser as I can't figure out how to revert or squash commits.
If Peter or Tom can cherry pick in the icarous.xml, we can later fix the comments in ardupilotmega.xml to note assigning of 42000-42099 to NASA/ICAROUS.  I'll raise an issue so we don't forget.

**Comment by peterbarker** [comment]
@auturgy I have squashed commits together, removed incorrect comment and whitespace.

Unfortunately, I have also adjusted pretty much every visible name in the PR.  The enumerations were no-brainers; they must include the name of the enumeration in every entry in the enumeration to avoid namespace conflicts.

I have also, however, prepended `ICAROUS_` to all names as well.  This is also for namespace reasons, and makes it match `uAvionix.xml`, which I believe to be a reasonable file to model things on.

Sadly, this will require code changes by the original authors.

Please check through my changes for consistency and that they are reasonable.

**Comment by tridge** [comment]
Looks good, I've pushed this with a squash, thanks for the work on this @peterbarker and @auturgy

**Pull request closed without merging**